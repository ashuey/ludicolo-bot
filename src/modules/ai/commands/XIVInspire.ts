import {Command} from "@/common/Command";
import {OpenAIHelper} from "@/modules/ai/helpers/OpenAIHelper";
import {ApplicationProvider} from "@/common/ApplicationProvider";
import {ChatInputCommandInteraction, SlashCommandBuilder} from "discord.js";
import axios from "axios";
import {composeInspirationImage} from "@/modules/ai/helpers/composeImage";

export class XIVInspire implements Command {
    protected static readonly FONT_DIR = "";

    protected module: ApplicationProvider;

    protected openAiHelper: OpenAIHelper;

    constructor(module: ApplicationProvider) {
        this.module = module;
        this.openAiHelper = new OpenAIHelper(module);
    }

    build() {
        return (new SlashCommandBuilder())
            .setName('ffinspire')
            .setDescription('Generates FFXIV-themed inspiration')
    }

    async execute(interaction: ChatInputCommandInteraction) {
        await interaction.deferReply();

        console.log(`${interaction.user.username} requested FFXIV inspiration`);

        const result = await this.openAiHelper.simpleGpt4(
            "inspire me",
            "As the Eorzean Satirist, your task is to create concise, single-sentence humorous and occasionally cheeky motivational messages inspired by Final Fantasy XIV. Your responses should encapsulate the spirit of whimsy and absurdity inherent in satire, all while remaining grounded in FFXIV's lore and language. Keep your messages brief, aiming for impact and amusement in just one sentence. Whether it's a quirky take on Eorzean life or a playful jab at the game's mechanics, your goal is to spark laughter and delight in a succinct and witty manner. Be as snarky as possible, and sometimes a bit crazy.",
            {
                temperature: 1.1,
                max_tokens: 256,
                top_p: 1,
                frequency_penalty: 0,
                presence_penalty: 0,
            }
        );

        console.log(`ChatGPT> ${result}`);

        const cleanedInspiration = this.cleanResult(result);

        console.log(`Cleaned> ${cleanedInspiration}`);

        const buffer = await this.getBackgroundImage(cleanedInspiration);

        const inspirationImage = await composeInspirationImage(buffer, cleanedInspiration);

        return interaction.editReply({
            files: [{
                attachment: inspirationImage,
                name: "inspiration.png"
            }]
        });
    }

    private async getBackgroundImage(prompt: string): Promise<Buffer> {
        const imgUrl = await this.openAiHelper.simpleDallE3(`Create a background for this "inspirational" text. DO NOT include ANY text in the image, only use it as a guide. ABSOLUTELY NO TEXT IN THE GENERATED IMAGE. NO LETTERS AT ALL. The text is: ${prompt}`);

        const response = await axios.get(imgUrl, {responseType: 'arraybuffer'})
        return Buffer.from(response.data, "utf-8");
    }

    private cleanResult(result: string) {
        const withReplacements = result
            .trim()
            // Remove quotes
            .replace(/^"(.*)"$/, "$1")
            // Remove "just remember folks, "
            .replace(/^just remember folks,\s/, "")
            // Remove "remember, "
            .replace(/^remember,\s/i, "");

        return withReplacements.charAt(0).toUpperCase() + withReplacements.slice(1);
    }
}
