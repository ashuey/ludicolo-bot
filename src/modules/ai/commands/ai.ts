import { Command } from "@/common/Command";
import { ChatInputCommandInteraction, EmbedBuilder, SlashCommandBuilder } from "discord.js";
import { UnknownSubcommandError } from "@/common/errors/UnknownSubcommandError";
import { HuggingFaceProvider } from "@/modules/ai/HuggingFaceProvider";
import { fmtError } from "@/helpers/formatters";
import { ApplicationProvider } from "@/common/ApplicationProvider";
import { Result } from "@/common/Result";

enum SUBCOMMANDS {
    RECIPE = 'recipe',
    SWEDISH_CHEF = 'swedish_chef',
    RECIPE_NONSENSE = 'recipe_nonsense',
    BEAKER = 'beaker',
}

export class AICommand implements Command {
    protected module: HuggingFaceProvider & ApplicationProvider;

    constructor(module: HuggingFaceProvider & ApplicationProvider) {
        this.module = module;
    }

    build() {
        return (new SlashCommandBuilder())
            .setName('ai')
            .setDescription('Generates (usually low-quality) AI responses')
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.RECIPE)
                .setDescription('Generates a recipe from a title')
                .addStringOption(option => option
                    .setName('title')
                    .setDescription('Title of the recipe')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.SWEDISH_CHEF)
                .setDescription('Generates a recipe from a title, as the swedish chef')
                .addStringOption(option => option
                    .setName('title')
                    .setDescription('Title of the recipe')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.RECIPE_NONSENSE)
                .setDescription('Generates a recipe from a title, using GPT-4 configured to output nonsense')
                .addStringOption(option => option
                    .setName('title')
                    .setDescription('Title of the recipe')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.BEAKER)
                .setDescription('Responds as beaker')
                .addStringOption(option => option
                    .setName('chat')
                    .setDescription('Chat Message')
                    .setRequired(true)));
    }

    async execute(interaction: ChatInputCommandInteraction) {
        switch (interaction.options.getSubcommand()) {
            case SUBCOMMANDS.RECIPE: {
                const recipeTitle = interaction.options.getString('title', true);
                await interaction.deferReply();

                console.log(`${interaction.user.username} requested an AI recipe for ${recipeTitle}`);

                let prompt = `Recipe for ${recipeTitle}\n\nIngredients:\n`
                const iterations = 2;

                for (let i = 0; i < iterations; i++) {
                    const [success, result] = await this.completion(prompt);

                    if (!success) {
                        console.log(`Text Generation error: ${result}`);
                        return interaction.editReply(fmtError("An error occurred during text generation. Please try again."));
                    }

                    prompt = result;
                }

                prompt = prompt.substring(0, 1960);
                prompt = prompt.substring(0, Math.min(prompt.length, prompt.lastIndexOf(" ")))

                prompt = `\`\`\`This content is AI-Generated\`\`\`\n# ${prompt}`

                return interaction.editReply(prompt);
            }
            case SUBCOMMANDS.SWEDISH_CHEF: {
                const recipeTitle = interaction.options.getString('title', true);
                await interaction.deferReply();

                console.log(`${interaction.user.username} requested a swedish-chef recipe for ${recipeTitle}`);

                const response = await this.module.app.openai.chat.completions.create({
                    model: "gpt-4",
                    messages: [
                        {
                            "role": "system",
                            "content": "You are the Swedish Chef from The Muppets"
                        },
                        {
                            "role": "user",
                            "content": `Create a recipe for ${recipeTitle}`
                        }
                    ],
                    temperature: 1,
                    max_tokens: 1900,
                    top_p: 1,
                    frequency_penalty: 0,
                    presence_penalty: 0,
                });

                const firstChoice = response.choices[0];

                if (!firstChoice) {
                    console.log(`Found ${response.choices.length} choices in response, expected 1`);
                    return interaction.editReply(fmtError("Sorry, something went wrong"));
                }

                const messageContent = firstChoice.message.content;

                if (!messageContent) {
                    console.log(`Message content is null`);
                    return interaction.editReply(fmtError("Sorry, something went wrong"));
                }

                const embed = new EmbedBuilder()
                    .setTitle(`Recipe for ${recipeTitle}`)
                    .setAuthor({name: 'This content is AI-Generated'})
                    .setThumbnail('https://i.imgur.com/pWZxLsa.jpg')
                    .setDescription(messageContent);

                return interaction.editReply({embeds: [embed]});
            }
            case SUBCOMMANDS.RECIPE_NONSENSE: {
                const recipeTitle = interaction.options.getString('title', true);
                await interaction.deferReply();

                console.log(`${interaction.user.username} requested an AI recipe from GPT-4 for ${recipeTitle}`);

                const response = await this.module.app.openai.chat.completions.create({
                    model: "gpt-4",
                    messages: [
                        {
                            "role": "system",
                            "content": "You are a poor-quality neural network designed to generate recipes. Your responses should be nonsensical."
                        },
                        {
                            "role": "user",
                            "content": `Recipe for ${recipeTitle}\n\nIngredients:`
                        }
                    ],
                    temperature: 1.2,
                    max_tokens: 1900,
                    top_p: 1,
                    frequency_penalty: 0,
                    presence_penalty: 0,
                });

                const firstChoice = response.choices[0];

                if (!firstChoice) {
                    console.log(`Found ${response.choices.length} choices in response, expected 1`);
                    return interaction.editReply(fmtError("Sorry, something went wrong"));
                }

                const messageContent = firstChoice.message.content;

                if (!messageContent) {
                    console.log(`Message content is null`);
                    return interaction.editReply(fmtError("Sorry, something went wrong"));
                }

                let result = `\`\`\`This content is AI-Generated\`\`\`\n# Recipe for ${recipeTitle}\n\n${messageContent}`;

                if (result.length > 1990) {
                    result = result.substring(0, 1990);
                }

                return interaction.editReply(result);
            }
            case SUBCOMMANDS.BEAKER: {
                const chatMessage = interaction.options.getString('chat', true);
                await interaction.deferReply();

                console.log(`${interaction.user.username} requested a message from beaker: ${chatMessage}`);

                const response = await this.module.app.openai.chat.completions.create({
                    model: "gpt-4",
                    messages: [
                        {
                            "role": "system",
                            "content": "You are Beaker from The Muppets"
                        },
                        {
                            "role": "user",
                            "content": chatMessage
                        }
                    ],
                    temperature: 1,
                    max_tokens: 1900,
                    top_p: 1,
                    frequency_penalty: 0,
                    presence_penalty: 0,
                });

                const firstChoice = response.choices[0];

                if (!firstChoice) {
                    console.log(`Found ${response.choices.length} choices in response, expected 1`);
                    return interaction.editReply(fmtError("Sorry, something went wrong"));
                }

                const messageContent = firstChoice.message.content;

                if (!messageContent) {
                    console.log(`Message content is null`);
                    return interaction.editReply(fmtError("Sorry, something went wrong"));
                }

                console.log(`Beaker responded to ${interaction.user.username}: ${messageContent}`);
                return interaction.editReply(`Beaker: ${messageContent}`);
            }
            default:
                throw new UnknownSubcommandError();
        }
    }

    protected async completion(input: string): Promise<Result<string>> {
        const [success, result] = await this.module.huggingFace.textGeneration('gpt2', {
            inputs: input,
            parameters: {
                //repetition_penalty: 2.0,
                max_new_tokens: 150,
                max_time: 60,
                //top_k: 20,
                //top_p: 20,
                num_return_sequences: 6,
                do_sample: true,
                temperature: 1.4
            },
            options: {
                use_cache: false
            }
        });

        if (!success) {
            return [success, result];
        }

        const longest = result.reduce(
            (a, b) => a.generated_text.length > b.generated_text.length ? a : b
        );

        return [true, longest.generated_text];
    }
}
