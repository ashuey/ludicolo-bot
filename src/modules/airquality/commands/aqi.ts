import { bold, ChatInputCommandInteraction, EmbedBuilder, SlashCommandBuilder } from "discord.js";
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import timezone from 'dayjs/plugin/timezone';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import { AirQualityColors } from "@/modules/airquality/airnow/colors";
import { AirNowProvider } from "@/modules/airquality/AirNowProvider";
import { Command } from "@/common/Command";
import { Forecast } from "@/modules/airquality/airnow/Forecast";
import { Observation } from "@/modules/airquality/airnow/Observation";
import { betterPollutantNames } from "@/modules/airquality/airnow/pollutants";
import {logger} from "@/logger";

dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(customParseFormat);

export class AQICommand implements Command {
    // TODO: Don't hardcode this
    static readonly DEFAULT_ZIP = 14614;

    protected module: AirNowProvider

    constructor(module: AirNowProvider) {
        this.module = module;
    }

    build() {
        return (new SlashCommandBuilder())
            .setName('aqi')
            .setDescription('Gets the current Air Quality Index for a location')
            .addIntegerOption(option =>
                option.setName('zip_code')
                    .setDescription('5-digit ZIP Code')
                    .setMinValue(10000)
                    .setMaxValue(99999))
    }

    async execute(interaction: ChatInputCommandInteraction) {
        const zip = interaction.options.getInteger('zip_code') ?? AQICommand.DEFAULT_ZIP;

        const response = (new EmbedBuilder())
            .setFooter({ text: "Data provided by AirNow.gov" });

        const current = await this.getCurrentObservations(zip);
        let now = dayjs();

        if (current) {
            now = now.tz(current.LocalTimeZone);
        }

        const [today, tomorrow] = await Promise.all([
            this.getForecast(zip, now),
            this.getForecast(zip, now.add(1, 'day')),
        ])

        let title = "Air Quality";
        let aqi = 'Unavailable';
        let primaryPollutant = 'Unavailable';
        let todayStr = 'Unavailable';
        let tomorrowStr = 'Unavailable';
        let currentCategory = 'Unavailable';

        if (current) {
            const color = AirQualityColors[current.Category.Number];

            // Add location to the title
            title = title + ` in ${current.ReportingArea}`;
            currentCategory = `${color ? `${color[1]} ` : ''} ${current.Category.Name}`

            // Get the timestamp of the observation
            const timestamp = dayjs.tz(
                `${current.DateObserved.trim()} ${current.HourObserved}`,
                'YYYY-MM-DD HH',
                current.LocalTimeZone
            ).toDate();

            const pollutantKey = current.ParameterName.toLowerCase().trim();

            aqi = String(current.AQI);
            primaryPollutant = betterPollutantNames[pollutantKey] ?? current.ParameterName;

            response
                .setTimestamp(timestamp)
                .setColor(color ? color[0] : null);
        }

        if (today) {
            const emoji = AirQualityColors[today.Category.Number];
            todayStr = `${emoji ? emoji[1] : ''} ${today.Category.Name}`.trim();
        }

        if (tomorrow) {
            const emoji = AirQualityColors[tomorrow.Category.Number];
            tomorrowStr = `${emoji ? emoji[1] : ''} ${tomorrow.Category.Name}`.trim();
        }

        response
            .setTitle(title)
            .setDescription(currentCategory)
            .addFields(
                {
                    name: 'Air Quality Index',
                    value: aqi,
                    inline: true,
                },
                {
                    name: 'Primary Pollutant',
                    value: primaryPollutant,
                    inline: true,
                },
                {
                    name: 'Forecast',
                    value: `${bold('Today:')} ${todayStr}\n${bold('Tomorrow:')} ${tomorrowStr}`,
                },
            );

        return interaction.reply({embeds: [response]});
    }

    protected async getCurrentObservations(zip: number): Promise<Observation|null> {
        const [obResult, observations] = await this.module.airNow.currentObservation(String(zip));

        if (!obResult) {
            logger.warn(`Error getting air quality observations: ${observations}`);

            return null;
        }

        let observation: Observation | null = null;

        observations.forEach(o => {
            if (observation === null || o.AQI > observation.AQI) {
                observation = o;
            }
        })

        return observation;
    }

    protected async getForecast(zip: number, date: dayjs.Dayjs): Promise<Forecast|null> {
        const dateString = date.format('YYYY-MM-DD');
        const [fResult, forecasts] = await this.module.airNow.forecasts(String(zip), dateString);

        if (!fResult) {
            logger.warn(`Error getting air quality forecasts for ${dateString}: ${forecasts}`);

            return null;
        }

        const forecastsForDate = forecasts.filter(f =>
            f.DateForecast.toLowerCase().trim() === dateString
        );

        let forecast: Forecast | null = null;

        forecastsForDate.forEach(f => {
            if (forecast === null || f.AQI > forecast.AQI) {
                forecast = f;
            }
        })

        return forecast;
    }
}
