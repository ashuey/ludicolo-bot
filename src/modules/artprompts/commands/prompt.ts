import { Command } from "@/common/Command";
import { ActionRowBuilder, ButtonBuilder, ChatInputCommandInteraction, SlashCommandBuilder, ButtonStyle } from "discord.js";
import { UnknownSubcommandError } from "@/common/errors/UnknownSubcommandError";
import { RuntimeError } from "@/common/errors/RuntimeError";
import { choose } from "@/helpers/random";
import Markov from 'markov-strings'
import {corpusSmall} from "@/modules/artprompts/data/corpus_small";
import {curatedPrompts} from "@/modules/artprompts/data/curated_prompts";

enum SUBCOMMANDS {
    RANDOM = 'random',
    AI_LOW = 'ai-low',
}

export class ArtPromptCommand implements Command {
    protected readonly markov: Markov;

    constructor() {
        this.markov = new Markov();
        this.markov.addData(corpusSmall);
    }
    build() {
        return (new SlashCommandBuilder())
            .setName('art-prompt')
            .setDescription('Gives a random art prompt')
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.RANDOM)
                .setDescription('Picks a random prompt from a list'))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.AI_LOW)
                .setDescription('Uses a low-quality AI to generate a prompt'));
    }

    async execute(interaction: ChatInputCommandInteraction) {
        switch (interaction.options.getSubcommand()) {
            case SUBCOMMANDS.RANDOM: {
                const prompt = choose(curatedPrompts);
                if (!prompt) {
                    throw new RuntimeError("Could not find random prompt");
                }
                return interaction.reply(`Art Prompt: ${prompt}`);
            }
            case SUBCOMMANDS.AI_LOW: {
                const prompt = this.markov.generate();

                const drawButton = new ButtonBuilder()
                    .setCustomId('com://art_prompts/draw_prompt')
                    .setLabel('Draw it for Me')
                    .setStyle(ButtonStyle.Primary);

                const row = new ActionRowBuilder<ButtonBuilder>()
                    .addComponents(drawButton);

                return interaction.reply({
                    content: `Art Prompt: ${prompt.string}`,
                    components: [row]
                });
            }
            default:
                throw new UnknownSubcommandError();
        }
    }
}
