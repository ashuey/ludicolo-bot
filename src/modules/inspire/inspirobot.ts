import axios, { AxiosInstance } from "axios";
import { Result } from "@/common/Result";

export class InspiroBot {
    static readonly ENDPOINT = "https://inspirobot.me/api?generate=true";

    protected readonly http: AxiosInstance;

    constructor() {
        this.http = axios.create({
            timeout: 2000,
            validateStatus: () => true,
        });
    }

    public async newImage(): Promise<Result<string>> {
        const result = await this.http.get<string>(InspiroBot.ENDPOINT);

        if (result.status < 200 || result.status >= 300) {
            return [false, new Error(`HTTP Error ${result.status}`)];
        }

        return [true, result.data];
    }
}

export const inspiroBot = new InspiroBot();
