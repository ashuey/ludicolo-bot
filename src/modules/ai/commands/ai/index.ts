import {ChatInputCommandInteraction, SlashCommandBuilder} from "discord.js";
import {Command} from "@/common/Command";
import {UnknownSubcommandError} from "@/common/errors/UnknownSubcommandError";
import {ApplicationProvider} from "@/common/ApplicationProvider";
import {OpenAIHelper} from "@/modules/ai/helpers/OpenAIHelper";
import {chefUriangerEmbed, swedishChefEmbed} from "@/modules/ai/commands/ai/embeds";
import {fmtAi, truncate} from "@/helpers/formatters";
import {logger} from "@/logger";
import {badWritingInputs} from "@/modules/ai/commands/ai/badWritingInputs";

enum SUBCOMMANDS {
    RECIPE = 'recipe',
    SWEDISH_CHEF = 'swedish_chef',
    BEAKER = 'beaker',
    URIANGER = 'urianger',
    CHEF_URIANGER = 'chef_urianger',
    BAD_WRITING = 'bad_writing',
    IMAGE = 'image',
    IMAGE_LEGACY = 'image_legacy',
}

export class AICommand implements Command {
    protected module: ApplicationProvider;

    protected openAiHelper: OpenAIHelper;

    constructor(module: ApplicationProvider) {
        this.module = module;
        this.openAiHelper = new OpenAIHelper(module);
    }

    build() {
        return (new SlashCommandBuilder())
            .setName('ai')
            .setDescription('Generates (usually low-quality) AI responses')
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.BAD_WRITING)
                .setDescription('Generates an atrocious opening sentence to the worst novel never written'))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.SWEDISH_CHEF)
                .setDescription('Generates a recipe from a title, as the swedish chef')
                .addStringOption(option => option
                    .setName('title')
                    .setDescription('Title of the recipe')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.RECIPE)
                .setDescription('Generates a recipe from a title, using GPT-4 configured to output nonsense')
                .addStringOption(option => option
                    .setName('title')
                    .setDescription('Title of the recipe')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.BEAKER)
                .setDescription('Responds as beaker')
                .addStringOption(option => option
                    .setName('chat')
                    .setDescription('Chat Message')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.URIANGER)
                .setDescription('Gives inspiration as Urianger'))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.CHEF_URIANGER)
                .setDescription('Generates a recipe from a title, as Urianger')
                .addStringOption(option => option
                    .setName('title')
                    .setDescription('Title of the recipe')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.IMAGE_LEGACY)
                .setDescription('Generates an image with Dall-E 2')
                .addStringOption(option => option
                    .setName('prompt')
                    .setDescription('Image prompt')
                    .setRequired(true)))
            .addSubcommand(cmd => cmd
                .setName(SUBCOMMANDS.IMAGE)
                .setDescription('Generates an image')
                .addStringOption(option => option
                    .setName('prompt')
                    .setDescription('Image prompt')
                    .setRequired(true)));
    }

    async execute(interaction: ChatInputCommandInteraction) {
        await interaction.deferReply();

        switch (interaction.options.getSubcommand()) {
            case SUBCOMMANDS.BAD_WRITING:
                return await this.badWritingSubcommand(interaction);
            case SUBCOMMANDS.SWEDISH_CHEF:
                return await this.swedishChefSubcommand(interaction);
            case SUBCOMMANDS.RECIPE:
                return await this.recipeSubcommand(interaction);
            case SUBCOMMANDS.BEAKER:
                return await this.beakerSubcommand(interaction);
            case SUBCOMMANDS.URIANGER:
                return await this.uriangerSubcommand(interaction);
            case SUBCOMMANDS.CHEF_URIANGER:
                return await this.chefUriangerSubcommand(interaction);
            case SUBCOMMANDS.IMAGE_LEGACY:
                return await this.imageLegacySubcommand(interaction);
            case SUBCOMMANDS.IMAGE:
                return await this.imageSubcommand(interaction);
            default:
                throw new UnknownSubcommandError();
        }
    }

    protected async badWritingSubcommand(interaction: ChatInputCommandInteraction) {
        const result = await this.openAiHelper.simpleGpt4(
            badWritingInputs.join("\n"),
            'WRITE MORE SENTENCES LIKE THESE',
            {
                model: "gpt-3.5-turbo",
                temperature: 1.17,
                max_tokens: 1024,
                top_p: 1,
                frequency_penalty: 0,
                presence_penalty: 0.88,
            }
        )

        const resultStr = result.split('\n')[0] ?? '';

        return interaction.editReply(truncate(fmtAi(resultStr)));
    }

    protected async swedishChefSubcommand(interaction: ChatInputCommandInteraction) {
        const recipeTitle = interaction.options.getString('title', true);

        logger.info(`${interaction.user.username} requested a swedish-chef recipe for ${recipeTitle}`);

        const messageContent = await this.openAiHelper.simpleGpt4(
            `Create a recipe for ${recipeTitle}`,
            "You are the Swedish Chef from The Muppets"
        );

        return interaction.editReply({content: fmtAi(""), embeds: [swedishChefEmbed(recipeTitle, messageContent)]});
    }

    protected async recipeSubcommand(interaction: ChatInputCommandInteraction) {
        const recipeTitle = interaction.options.getString('title', true);

        logger.info(`${interaction.user.username} requested an AI recipe from GPT-4 for ${recipeTitle}`);

        const messageContent = await this.openAiHelper.simpleGpt4(
            `Recipe for ${recipeTitle}\n\nIngredients:`,
            "You are a poor-quality neural network designed to generate recipes. Your responses should be nonsensical."
        );

        return interaction.editReply(truncate(fmtAi(`# Recipe for ${recipeTitle}\n\n${messageContent}`)));
    }

    protected async beakerSubcommand(interaction: ChatInputCommandInteraction) {
        const chatMessage = interaction.options.getString('chat', true);

        logger.info(`${interaction.user.username} requested a message from beaker: ${chatMessage}`);

        const response = await this.openAiHelper.simpleGpt4(chatMessage, "You are Beaker from The Muppets");

        return interaction.editReply(truncate(fmtAi(`Beaker: ${response}`)));
    }

    protected async uriangerSubcommand(interaction: ChatInputCommandInteraction) {
        const response = await this.openAiHelper.simpleGpt4("Inspire me, as if it were a motivational poster. No more than one or two sentences.", "You are Urianger from Final Fantasy XIV");

        return interaction.editReply(truncate(fmtAi(`Urianger: ${response}`)));
    }

    protected async chefUriangerSubcommand(interaction: ChatInputCommandInteraction) {
        const recipeTitle = interaction.options.getString('title', true);

        logger.info(`${interaction.user.username} requested an Urianger recipe for ${recipeTitle}`);

        const messageContent = await this.openAiHelper.simpleGpt4(
            `Create a recipe for ${recipeTitle}`,
            "You are Urianger from Final Fantasy XIV"
        );

        return interaction.editReply({content: fmtAi(""), embeds: [chefUriangerEmbed(recipeTitle, messageContent)]});
    }

    protected async imageLegacySubcommand(interaction: ChatInputCommandInteraction) {
        const prompt = interaction.options.getString('prompt', true);

        const images = await this.openAiHelper.simpleDallE2(prompt);

        return interaction.editReply({
            content: fmtAi(`Prompt: ${prompt}`),
            files: images.map((imgUrl, idx) => ({
                attachment: imgUrl,
                name: `image${idx}.png`
            }))
        })
    }

    protected async imageSubcommand(interaction: ChatInputCommandInteraction) {
        const prompt = interaction.options.getString('prompt', true);

        const imgUrl = await this.openAiHelper.simpleDallE3(`Draw this: ${prompt}`);

        return interaction.editReply({
            content: fmtAi(`Prompt: ${prompt}`),
            files: [{
                attachment: imgUrl,
                name: "inspiration.png"
            }]
        })
    }
}
