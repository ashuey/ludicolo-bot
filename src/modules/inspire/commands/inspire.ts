import { Command } from "@/common/Command";
import { ChatInputCommandInteraction, SlashCommandBuilder } from "discord.js";
import { inspiroBot } from "@/modules/inspire/inspirobot";

export class InspireCommand implements Command {
    build() {
        return (new SlashCommandBuilder())
            .setName('inspire')
            .setDescription('Inspires you with an AI-generated quote from InspiroBot');
    }

    async execute(interaction: ChatInputCommandInteraction) {
        const [success, result] = await inspiroBot.newImage();

        if (!success) {
            return interaction.reply({
                content: "Sorry, we're running low on inspiration. (An error occurred)",
                ephemeral: true
            });
        }

        console.debug(`Bestowing inspiration upon ${interaction.user.username}`)

        return interaction.reply({ files: [result] });
    }
}
