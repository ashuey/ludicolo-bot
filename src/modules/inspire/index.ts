import { Module } from "@/common/Module";
import { InspireCommand } from "@/modules/inspire/commands/inspire";

export class InspireModule implements Module {
    readonly name = 'inspire';

    readonly commands = [
        new InspireCommand()
    ];

    readonly migrations = [];
}
