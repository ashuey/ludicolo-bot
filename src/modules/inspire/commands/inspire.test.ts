import {InspireCommand} from "@/modules/inspire/commands/inspire";
import {inspiroBot} from "@/modules/inspire/inspirobot";

describe('InspireCommand', () => {
    it('builds the inspire command correctly', () => {
        const command = (new InspireCommand()).build();
        expect(command.toJSON().name).toBe('inspire');
    });

    it("test execute of InspireCommand for success", async () => {
        const inspireCommand = new InspireCommand();
        const mockInteraction = {
            reply: jest.fn(() => Promise.resolve()),
            user: {
                username: 'TEST-USERNAME'
            }
        };
        jest.spyOn(inspiroBot, 'newImage')
            .mockImplementation(() => Promise.resolve([true, 'SAMPLE-IMAGE-URL']));

        await inspireCommand.execute(mockInteraction as never);

        expect(mockInteraction.reply.mock.calls).toMatchSnapshot();
    });

    it("test execute of InspireCommand for failure", async () => {
        const inspireCommand = new InspireCommand();
        const mockInteraction = {
            reply: jest.fn(() => Promise.resolve()),
        };
        jest.spyOn(inspiroBot, 'newImage')
            .mockImplementation(() => Promise.resolve([false, new Error()]));

        await inspireCommand.execute(mockInteraction as never);

        expect(mockInteraction.reply.mock.calls).toMatchSnapshot();
    });
});
