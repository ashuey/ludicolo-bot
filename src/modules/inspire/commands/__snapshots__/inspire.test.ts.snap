// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspireCommand test execute of InspireCommand for failure 1`] = `
[
  [
    {
      "content": "Sorry, we're running low on inspiration. (An error occurred)",
      "ephemeral": true,
    },
  ],
]
`;

exports[`InspireCommand test execute of InspireCommand for success 1`] = `
[
  [
    {
      "files": [
        "SAMPLE-IMAGE-URL",
      ],
    },
  ],
]
`;
